### PR TITLE
Tools: rewrite delete can delete series only if it matches all matchers

### LINKED ISSUE #3885: tools bucket rewrite: Impossible to have multiple matchers?
**Thanos, Prometheus and Golang version used**:
thanos v0.19.0-rc.0

**Object Storage Provider**:
S3

**What happened**:
I tried to rewrite some blocks in dry-run for now and I am not happy with the output...
I tried to add multiple matchers (either by multiple arrays or multiple label matchers) and for both solution, it is doing an "OR" matcher and not an "AND" matcher like prometheus:

For example if I try:
```
- matchers: \"{job=\\\"job-to-delete\\\",__name__=\\\"up\\\"}\"
```
It will delete ALL timeseries of the job "job-to-delete" AND all metrics "up" from all other jobs as well.

Same if I try something like:
```
- matchers: \"{job=\\\"job-to-delete\\\"}\"
- matchers: \"{__name__=\\\"up\\\"}\"
```

**What you expected to happen**:
I would like only the "up" metrics of the job "job-to-delete" to be removed

**How to reproduce it (as minimally and precisely as possible)**:
Dry run a deletion with a matcher for a common metric (up) and a job:
```
thanos tools bucket rewrite --id=01EZW9YDX86ZGANKC5VY3E8DHZ --objstore.config-file=bucket-prd.yaml --rewrite.to-delete-config="
- matchers: \"{__name__=~\\\"up\\\",job=~\\\"anyjob\\\"}\"
" --dry-run --log.level=debug
```

`cat` the change.log and see that all "up" metrics are deleted wether or not part of "anyjob", and some "non up" metrics are deleted from the job "anyjob".

**Anything else we need to know**:
I would also expect something like:
```
- matchers: \"{job=\\\"job-to-delete\\\",job=\\\"another-job\\\"}\"
```
to never match and delete anything.

### PR DESCRIPTION
**Pull request by yeya24**
* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

Fixes #3885.
For all matchers in a single deletion request, we should use `AND` between matchers instead of `OR`.

## Verification

One test case added.

### CODE REVIEW

File: pkg/compactv2/compactor_test.go [hunk 1]
```diff
@@ -274,6 +274,38 @@ func TestCompactor_WriteSeries_e2e(t *testing.T) {
 				NumChunks:  2,
 			},
 		},
+		{
+			name: "1 blocks + delete modifier, deletion request contains multiple matchers, delete second series",
```
Comment (squat) [inline pkg/compactv2/compactor_test.go:278]: :+1:

**Review by squat — APPROVED** [review]
good find :+1:

**Review by bwplotka — APPROVED** [review]
Thanks! 🤗

**Comment by bwplotka** [comment]
Should go on v0.19.0-rc.0?

**Comment by squat** [comment]
Yeah i think so. Maybe this actually should have been made against that branch

**Comment by bwplotka** [comment]
That's fine, let's replicate commit

**Comment by bwplotka** [comment]
https://github.com/thanos-io/thanos/pull/3890

**Pull request merged**